File: jevpilot/jev.py
"""Thin HTTP client for TypeSafe's System One endpoint (Jev)."""
import time

import requests

ENDPOINT = "https://api.typesafe.ai/v1/systemone"
MODEL = "jev-latest"
MAX_CHOICE_OPTIONS = 255


class JevError(RuntimeError):
    pass


class Jev:
    def __init__(self, api_key: str, model: str = MODEL):
        self.api_key = api_key.strip()
        self.model = model
        self.session = requests.Session()
        self.tokens_in = 0

    def ask(self, state, questions: dict, retries: int = 3) -> tuple[dict, float]:
        """Send one System One request. Returns (answers, seconds)."""
        if not self.api_key:
            raise JevError("Kein API-Key gesetzt (Einstellungen).")
        body = {"model": self.model, "state": state, "questions": questions}
        headers = {"Authorization": f"Bearer {self.api_key}"}
        delay = 1.0
        for attempt in range(retries + 1):
            t0 = time.perf_counter()
            try:
                r = self.session.post(ENDPOINT, json=body, headers=headers, timeout=30)
            except requests.RequestException as e:
                if attempt == retries:
                    raise JevError(f"Netzwerkfehler: {e}") from e
                time.sleep(delay)
                delay *= 2
                continue
            dt = time.perf_counter() - t0
            if r.status_code == 200:
                data = r.json()
                self.tokens_in += data.get("usage", {}).get("input_tokens", 0)
                return data["answers"], dt
            if r.status_code in (429, 500, 502, 503, 504) and attempt < retries:
                time.sleep(delay)
                delay *= 2
                continue
            if r.status_code in (401, 403):
                raise JevError("API-Key ungültig oder ohne Rechte (HTTP %d)." % r.status_code)
            raise JevError(f"HTTP {r.status_code}: {r.text[:300]}")
        raise JevError("Keine Antwort von Jev.")


def choice(instructions: str, options: dict) -> dict:
    return {"type": "choice", "instructions": instructions, "criteria": options}


def noul(instructions: str) -> dict:
    return {"type": "noul", "instructions": instructions}

File: android/app/src/main/java/com/leo/jevpilot/Jev.kt
package com.leo.jevpilot

import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

class JevError(msg: String) : Exception(msg)

/** Thin client for TypeSafe's System One endpoint (Jev), same as jevpilot/jev.py on the PC. */
class Jev(private val apiKey: String) {
    companion object {
        const val ENDPOINT = "https://api.typesafe.ai/v1/systemone"
        const val MODEL = "jev-latest"
        const val MAX_CHOICE_OPTIONS = 255

        fun choice(instructions: String, options: Map<String, String>): JSONObject =
            JSONObject().put("type", "choice").put("instructions", instructions)
                .put("criteria", JSONObject(options))

        fun noul(instructions: String): JSONObject =
            JSONObject().put("type", "noul").put("instructions", instructions)
    }

    /** Returns (answers, seconds). */
    fun ask(state: Any, questions: JSONObject, retries: Int = 3): Pair<JSONObject, Double> {
        if (apiKey.isBlank()) throw JevError("No API key set (Settings).")
        val body = JSONObject().put("model", MODEL).put("state", state).put("questions", questions)
            .toString().toByteArray()
        var delay = 1000L
        for (attempt in 0..retries) {
            val t0 = System.nanoTime()
            val conn = URL(ENDPOINT).openConnection() as HttpURLConnection
            try {
                conn.requestMethod = "POST"
                conn.connectTimeout = 10_000
                conn.readTimeout = 30_000
                conn.doOutput = true
                conn.setRequestProperty("Authorization", "Bearer ${apiKey.trim()}")
                conn.setRequestProperty("Content-Type", "application/json")
                conn.outputStream.use { it.write(body) }
                val code = conn.responseCode
                val dt = (System.nanoTime() - t0) / 1e9
                if (code == 200) {
                    val data = JSONObject(conn.inputStream.bufferedReader().readText())
                    return data.getJSONObject("answers") to dt
                }
                val err = conn.errorStream?.bufferedReader()?.readText().orEmpty()
                if (code in listOf(429, 500, 502, 503, 504) && attempt < retries) {
                    Thread.sleep(delay); delay *= 2; continue
                }
                if (code == 401 || code == 403) throw JevError("API key invalid or not permitted (HTTP $code).")
                throw JevError("HTTP $code: ${err.take(300)}")
            } catch (e: IOException) {
                if (attempt == retries) throw JevError("Network error: ${e.message}")
                Thread.sleep(delay); delay *= 2
            } finally {
                conn.disconnect()
            }
        }
        throw JevError("No answer from Jev.")
    }
}
